### PR TITLE
fix(pam/integration-tests): Improve reliability of PAM integration tests and actually run them in race mode

### PR DESCRIPTION
**Pull request by 3v1n0**
Since we rely a lot on sleeping on the integration tests, due to the fact we can't yet go with https://github.com/charmbracelet/vhs/pull/257, we need to be able to tune this value depending on the context we're running the tests.

In general sanitizers slows down the runtime quite a bit (especially the thread one), so make possible to use dynamic sleep times values depending on variables that we adjust depending on the context we're running in.
Tests can now slowed up/down using `AUTHD_TESTS_SLEEP_MULTIPLIER` variable too (this can be used for helping local testing too, e.g. personally I can reliably run the tests faster in my machine - with no failures - using `AUTHD_TESTS_SLEEP_MULTIPLIER=0.3`; but also it helps to quickly "fix" slower machines).

Tune the tape files so that we don't miss some changing contents (as in the MFA/QrCode tests) and make the example broker to be a bit lazier.

Then, I noticed that we were not actually running the integration tests in `-race` mode, so fix this. Indeed this implies slower tests, but at least now we're fully checking for races both the daemon and the client.

This worked fine in various builds both in my fork and in a private repo fork I did where the builders are way slower than the public ones.

---

@adombeck please check if this improves the results in your local environment too, where you had failures

UDENG-4793

### CODE REVIEW

File: examplebroker/broker.go [hunk 1]
```diff
@@ -549,82 +570,93 @@ func (b *Broker) IsAuthenticated(ctx context.Context, sessionID, authenticationD
...
 		}
 
 	case "phoneack1":
 		// TODO: should this be an error rather (not expected data from the PAM module?
 		if authData["wait"] != "true" {
-			return AuthDenied, `{"message": "phoneack1 should have wait set to true"}`, nil
+			return AuthDenied, `{"message": "phoneack1 should have wait set to true"}`
 		}
-		// Send notification to phone1 and wait on server signal to return if OK or not
+		// Send notification to phone1 and wait on server signal to return if OK o
```
Comment (denisonbarbosa) [inline examplebroker/broker.go:619]: Seems like you cut a part of the comment here

File: pam/integration-tests/cli_test.go [hunk 2]
```diff
@@ -44,10 +40,10 @@ func TestCLIAuthenticate(t *testing.T) {
 		"Authenticate user with mfa":                                           {tape: "mfa_auth"},
 		"Authenticate user with form mode with button":                         {tape: "form_with_button"},
 		"Authenticate user with qr code":                                       {tape: "qr_code", pamUser: "user-integration-qr-code"},
-		"Authenticate user with qr code in a TTY":                              {tape: "qr_code", pamUser: "user-integration-qr-code-tty", termEnv: "linux"},
-		"Authenticate user with qr code in a TTY session":                      {tape: "qr_code", pamUser: "user-integration-qr-code-tty-session", termEnv: "xterm-256color", sessionEnv: "tty"},
-		"Authenticate user with qr code in screen":                             {tape: "qr_code", pamUser: "user-integration-qr-code-screen", termEnv: "screen"},
-		"Authenticate user with qr code after many regenerations":              {tape: "qr_code_quick_regenerate"},
+		"Authenticate user with qr code in a TTY":                              {tape: "qr_code", tapeSettings: []tapeSetting{{"Height", 650}}, pamUser: "user-integration-qr-code-tty", termEnv: "linux"},
```
Comment (denisonbarbosa) [inline pam/integration-tests/cli_test.go:43]: Maybe we can define the settings "keys" as consts in the helper file?

**Review by denisonbarbosa — APPROVED** [review]
This is a big one, although the review is quite small. I have some nitpicks, but nothing blocking so I'll leave my approval already. Nice work! Hopefully our tests get more stable now!

It's so frustrating to see how much we need to work around because we don't have the "wait for output" feature in vhs yet, but well...  It's not something we can control, right? 🥲

File: pam/integration-tests/cli_test.go (hunk 2 above)
Comment (3v1n0) [reply pam/integration-tests/cli_test.go:43]: You mean having a list of valid keys or define the defaults?

Comment (3v1n0) [reply pam/integration-tests/cli_test.go:43]: AH... ok I guess I got what you mean.

**Comment by adombeck** [comment]
> @adombeck please check if this improves the results in your local environment too, where you had failures

when I run it without setting `AUTHD_TESTS_SLEEP_MULTIPLIER` it's still flaky. It passed in the first run but in a second run it failed with the usual golden file mismatch error:

```
    --- FAIL: TestNativeAuthenticate/Authenticate_user_with_mfa (28.09s)
        native_test.go:126: 
            	Error Trace:	/home/adombeck/projects/authd/pam/integration-tests/native_test.go:126
            	Error:      	Not equal: [...]
            	Test:       	TestNativeAuthenticate/Authenticate_user_with_mfa
            	Messages:   	Output of tape "mfa_auth" does not match golden file
[...]
FAIL
FAIL	github.com/ubuntu/authd/pam/integration-tests	152.044s
```

I then tried running it with `AUTHD_TESTS_SLEEP_MULTIPLIER=0.3`. After about 3 minutes, it started using up all the 42 GiB of RAM on my system + the 64 GiB of swap and eventually caused my system to freeze.

I'll now try with `AUTHD_TESTS_SLEEP_MULTIPLIER=1.5` and report back.

**Comment by 3v1n0** [comment]
> FAIL: TestNativeAuthenticate/Authenticate_user_with_mfa (28.09s)

Do you have the generated golden file from that log?

It should be in `/tmp/authd-test-artifacts/` if you didn't preserve the log diff itself

**Comment by adombeck** [comment]
> I can try to reproduce the error and provide you the complete log if that helps.

Here is another error from running without `AUTHD_TESTS_SLEEP_MULTIPLIER`:

```
    --- FAIL: TestNativeAuthenticate/Exit_if_user_is_not_pre-checked_on_ssh_service (2.98s)
        native_test.go:122: 
            	Error Trace:	/home/adombeck/projects/authd/pam/integration-tests/native_test.go:122
            	Error:      	Received unexpected error:
            	            	exit status 1
            	Test:       	TestNativeAuthenticate/Exit_if_user_is_not_pre-checked_on_ssh_service
            	Messages:   	Failed to run tape "local_ssh": exit status 1: File: /tmp/TestNativeAuthenticateExit_if_user_is_not_pre-checked_on_ssh_service2432860924/001/local_ssh
            	            	Host your GIF on vhs.charm.sh: vhs publish <file>.gif
            	            	could not launch browser: [launcher] Failed to get the debug url: Failed to open /dev/null                             
            	            	[280538:280538:0930/164144.200144:FATAL:zygote_host_impl_linux.cc(207)] Check failed: . : No such file or directory (2)
            	            	[0930/164144.206417:ERROR:ptracer.cc(605)] ptrace: Input/output error (5)                                              
            	            	[0930/164144.330607:ERROR:elf_dynamic_array_reader.h(64)] tag not found                                                
            	            	                                                                                                                       
            	            	recording failed
```

Full test logs: [gotest.log](https://github.com/user-attachments/files/17191891/gotest.log)

**Comment by 3v1n0** [comment]
Ok the `Authenticate_user_and_add_it_to_local_group` issues are not due to integration tests but I think we had race due to the fact we may try to rely on filesystem content that may not have been yet updated or vice versa, https://github.com/ubuntu/authd/pull/560/commits/6d3afdd1d2b211a57fd035f516195fe44efe9922 should help with it.

**Comment by 3v1n0** [comment]
> --- FAIL: TestNativeAuthenticate/Exit_if_user_is_not_pre-checked_on_ssh_service (2.98s)

Not our bug: vhs (or the called chrome/chromium) doesn't seem to start, so not something we can fix.

File: pam/integration-tests/cli_test.go (hunk 2 above)
Comment (3v1n0) [reply pam/integration-tests/cli_test.go]: I'm using `const` values now, see if that's what you meant please :)

**Comment by adombeck** [comment]
Another failure:

```
    --- FAIL: TestNativeAuthenticate/Authenticate_user_on_ssh_service_with_custom_name_and_connection_env (3.51s)
        native_test.go:122: 
            	Error Trace:	/home/adombeck/projects/authd/pam/integration-tests/native_test.go:122
            	Error:      	Received unexpected error:
            	            	exit status 2
            	Test:       	TestNativeAuthenticate/Authenticate_user_on_ssh_service_with_custom_name_and_connection_env
            	Messages:   	Failed to run tape "simple_ssh_auth": exit status 2: File: /tmp/TestNativeAuthenticateAuthenticate_user_on_ssh_service_with_custom_name_and_connection_env3717920695/001/simple_ssh_auth
            	            	Host your GIF on vhs.charm.sh: vhs publish <file>.gif
            	            	Output .txt simple_ssh_auth.txt
            	            	Output .gif simple_ssh_auth.gif
            	            	Set Width 800
            	            	Set Height 500
            	            	Set FontFamily Monospace
            	            	Set FontSize 13
            	            	Set Padding 0
            	            	Set Margin 0
            	            	Set Shell bash
            	            	panic: write tcp 127.0.0.1:39266->127.0.0.1:36905: use of closed network connection
            	            	
            	            	goroutine 1 [running]:
            	            	github.com/go-rod/rod/lib/utils.init.func2({0xf65600?, 0xc0003dabe0?})
            	            		/home/adombeck/go/pkg/mod/github.com/go-rod/rod@v0.114.8/lib/utils/utils.go:69 +0x1d
            	            	main.(*VHS).Start.New.(*Browser).WithPanic.genE.func1({0xc00004bb40?, 0xc0003da960?, 0x14f9f00?})
            	            		/home/adombeck/go/pkg/mod/github.com/go-rod/rod@v0.114.8/must.go:36 +0x62
            	            	github.com/go-rod/rod.(*Page).MustSetViewport(0xc00023e160, 0x100006ff1f0?, 0x724b27282f88?, 0x10?, 0xb8?)
            	            		/home/adombeck/go/pkg/mod/github.com/go-rod/rod@v0.114.8/must.go:313 +0xe2
            	            	main.(*VHS).Setup(0xc00070e7e0)
            	            		/home/adombeck/go/pkg/mod/github.com/charmbracelet/vhs@v0.7.2/vhs.go:166 +0x7f
            	            	main.Evaluate({0x14ff778, 0xc0003e5400}, {0xc000264000, 0x19a}, {0x14f95c0, 0xc00008a058}, {0xc0006ffc30, 0x1, 0x0?})
            	            		/home/adombeck/go/pkg/mod/github.com/charmbracelet/vhs@v0.7.2/evaluator.go:79 +0x8f1
            	            	main.init.func6(0x1c0d4c0, {0xc0005de110, 0x1, 0x101e816?})
            	            		/home/adombeck/go/pkg/mod/github.com/charmbracelet/vhs@v0.7.2/main.go:99 +0x44e
            	            	github.com/spf13/cobra.(*Command).execute(0x1c0d4c0, {0xc000036050, 0x1, 0x1})
            	            		/home/adombeck/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:983 +0xaca
            	            	github.com/spf13/cobra.(*Command).ExecuteC(0x1c0d4c0)
            	            		/home/adombeck/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
            	            	github.com/spf13/cobra.(*Command).Execute(...)
            	            		/home/adombeck/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1039
            	            	github.com/spf13/cobra.(*Command).ExecuteContext(...)
            	            		/home/adombeck/go/pkg/mod/github.com/spf13/cobra@v1.8.0/command.go:1032
            	            	main.main()
            	            		/home/adombeck/go/pkg/mod/github.com/charmbracelet/vhs@v0.7.2/main.go:250 +0xc9
```

Full test logs: [gotest.log](https://github.com/user-attachments/files/17193091/gotest.log)

**Comment by 3v1n0** [comment]
@adombeck looks like another vhs upstream issue, please report it there.

**Comment by 3v1n0** [comment]
@adombeck, @denisonbarbosa: If there are no further blockers I'd proceed merging this so that I'm not blocked in further cleanups branch I've + other tests.

This seems more stable in CI and locally to me, but if there further failures I'm happy to analyze them (in general they require just some extra sleeps here and there, sadly)

**Pull request merged**